adoc/extensions/sycl_khr_free_function_commands.adoc

```diff
@@ -0,0 +1,967 @@
+  myQueue.wait();
+
+  std::cout << std::endl << "Result:" << std::endl;
+  for (size_t i = 0; i < N; i++) {
+    for (size_t j = 0; j < M; j++) {
+      size_t index = i * M + j;
+      // Compare the result to the analytic value.
+      if (c[index] != i * (2 + 2014) + j * (1 + 42)) {
+        std::cout << "Wrong value " << c[index] << " on element " << i << " "
+                  << j << std::endl;
+        exit(-1);
+      }
+    }
+  }
+
+  std::cout << "Good computation!" << std::endl;
+  return 0;
+}
+----
+
+== Requirements
+
+An instance of the [code]#requirements# class template encapsulates all of the
+requirements for scheduling a command.
+
+[source,role=synopsis]
+----
+namespace sycl::khr {
+
+template <typename T>
+struct is_requirement;
+
+template <typename T>
+static constexpr bool is_requirement_v = is_requirement<T>::value_type;
+
+template <typename... Requirements>
+class requirements {
+public:
+  // Only available if: all Requirements satisfy is_requirement_v
+  requirements(Requirements... values);
+
+  void register-accessors(handler& h) const;     // exposition only
+  void register-events(handler& h) const;        // exposition only
+  void register-kernel-bundle(handler& h) const; // exposition only
+  bool has-tracking() const;                     // exposition only
+};
+
+class tracking {
+public:
+  tracking(bool enabled=true);
+};
+
+}
+----
+
+Each instance of a type listed below defines a specific scheduling requirement.
+For each type, the [code]#is_requirement# type trait is specialized such that
+[code]#is_requirement_v<T># returns [code]#true#.
+
+* [code]#event#: The command must not begin executing until the event is
+  complete.
+
+* [code]#std::vector<event>#: The command must not begin executing until all
+  events in the vector are complete.
+
+* [code]#accessor#: The command must not begin executing until the
+  [code]#buffer# associated with the [code]#accessor# can be accessed in a
+  manner compatible with the specified [code]#access_mode#.
+
+* [code]#tracking#: The command must be submitted such that its state can be
+  tracked via an [code]#event# when the [code]#tracking# object is constructed
+  with an [code]#enabled# value of [code]#true#.
+
+* [code]#kernel_bundle<bundle_state::executable>#: The command must be submitted
+  using a <<device image>> from the kernel bundle.
+
+'''
+
+.[apititle]#Default constructor#
+[source,role=synopsis,id=api:requirements-ctor]
+----
+template <typename... Requirements>
+requirements(Requirements... values);
+----
+
+_Constraints_:
+* [code]#is_requirement_v# returns [code]#true# for each type in
+  [code]#Requirements#; and
+* [code]#Requirements# contains at most one
+  [code]#kernel_bundle<bundle_state::executable>#.
+
+_Effects_: Constructs a [code]#requirements# object representing the set of
+requirements specified via the [code]#values# parameter pack.
+
+_Remarks_: Unless otherwise specified, if an instance of a requirement appears
```

Comment:
When we refer to an instance appearing more than once, does this mean objects which compare equals? So for example SYCL classes with reference semantics.

Reply:
Yes, I think so.  As far as I can tell, it's legal to call `h.require(acc)` twice with the same accessor, or to call `h.depends_on(e)` with the same event multiple times.  I don't see a reason to disallow that behavior here.

Do you think we need to clarify here what it means for requirements to be equivalent?